File: demo/protocol.py
"""Binary framing for OmniTalker streaming S2ST WebSocket demo.

Frame layout: ``[type:u8][payload...]``

| type | Direction | Payload |
|------|-----------|---------|
| 0x00 | C↔S       | UTF-8 JSON (config / ready / error / done) |
| 0x01 | C→S       | PCM int16 LE mono @ 16 kHz |
| 0x01 | S→C       | PCM int16 LE mono @ 24 kHz |
| 0x02 | S→C       | UTF-8 translation text chunk |
| 0x03 | C→S       | EOS (empty payload) |
"""

from __future__ import annotations

import json
from typing import Any

TYPE_JSON = 0x00
TYPE_PCM = 0x01
TYPE_TEXT = 0x02
TYPE_EOS = 0x03

SAMPLE_RATE_IN = 16_000
SAMPLE_RATE_OUT = 24_000


def pack_frame(frame_type: int, payload: bytes = b"") -> bytes:
    if not (0 <= frame_type <= 255):
        raise ValueError(f"invalid frame_type={frame_type}")
    return bytes((frame_type,)) + payload


def unpack_frame(message: bytes) -> tuple[int, bytes]:
    if not message:
        raise ValueError("empty websocket message")
    return message[0], message[1:]


def pack_json(obj: dict[str, Any]) -> bytes:
    return pack_frame(TYPE_JSON, json.dumps(obj, ensure_ascii=False).encode("utf-8"))


def unpack_json(payload: bytes) -> dict[str, Any]:
    return json.loads(payload.decode("utf-8"))


def pack_pcm_i16(pcm: bytes) -> bytes:
    return pack_frame(TYPE_PCM, pcm)


def pack_text(text: str) -> bytes:
    return pack_frame(TYPE_TEXT, text.encode("utf-8"))


def pack_eos() -> bytes:
    return pack_frame(TYPE_EOS)

File: demo/chrome_extension/protocol.js
/** Binary framing shared with demo/protocol.py */

export const TYPE_JSON = 0x00;
export const TYPE_PCM = 0x01;
export const TYPE_TEXT = 0x02;
export const TYPE_EOS = 0x03;
export const SAMPLE_RATE_IN = 16000;
export const SAMPLE_RATE_OUT = 24000;

export function packFrame(type, payloadBytes = new Uint8Array(0)) {
  const out = new Uint8Array(1 + payloadBytes.byteLength);
  out[0] = type;
  out.set(payloadBytes, 1);
  return out.buffer;
}

export function unpackFrame(arrayBuffer) {
  const u8 = new Uint8Array(arrayBuffer);
  if (u8.byteLength < 1) throw new Error("empty websocket message");
  return { type: u8[0], payload: u8.subarray(1) };
}

export function packJson(obj) {
  const enc = new TextEncoder().encode(JSON.stringify(obj));
  return packFrame(TYPE_JSON, enc);
}

export function unpackJson(payload) {
  return JSON.parse(new TextDecoder().decode(payload));
}

export function packPcmI16(int16Array) {
  return packFrame(TYPE_PCM, new Uint8Array(int16Array.buffer, int16Array.byteOffset, int16Array.byteLength));
}

export function packEos() {
  return packFrame(TYPE_EOS);
}

/** Float32 mono [-1,1] → Int16Array */
export function floatToInt16(float32) {
  const out = new Int16Array(float32.length);
  for (let i = 0; i < float32.length; i++) {
    const x = Math.max(-1, Math.min(1, float32[i]));
    out[i] = (x * 32767) | 0;
  }
  return out;
}

/** Simple average downsample float32 mono. */
export function downsample(float32, fromRate, toRate) {
  if (fromRate === toRate) return float32;
  const ratio = fromRate / toRate;
  const outLen = Math.floor(float32.length / ratio);
  const out = new Float32Array(outLen);
  for (let i = 0; i < outLen; i++) {
    const start = Math.floor(i * ratio);
    const end = Math.min(float32.length, Math.floor((i + 1) * ratio));
    let sum = 0;
    for (let j = start; j < end; j++) sum += float32[j];
    out[i] = sum / Math.max(1, end - start);
  }
  return out;
}
